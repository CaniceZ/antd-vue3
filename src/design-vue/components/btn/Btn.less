.ygp-btn {
  &.ant-btn[disabled] {
    color: #c8c9cc;
    border-color: #ebedf0;
  }

  // .ant-btn:focus {
  //   background: #eaeaea;
  //   border: 1px solid #dcdee0;
  //   color: #323233;
  // }

  &.ant-btn.ant-btn-loading {
    .anticon-loading {
      svg {
        color: #fe5f23;
      }
    }
  }

  &.ant-btn-primary.ant-btn-loading {
    .anticon-loading {
      svg {
        color: #fff;
      }
    }
  }

  &.ant-btn-primary:focus {
    background-color: #e94000;
    border-color: #e94000;
  }

  &.ant-btn-link[disabled] {
    border-color: transparent;
  }

  &.ant-btn-dangerous.ant-btn-link:hover,
  &.ant-btn-dangerous.ant-btn-link:focus {
    // color: #e02a24;
    // background: #f2f3f5;
    background: #e02a24;
    color: white;
  }

  &.ant-btn-link {
    padding-left: 4px;
    padding-right: 4px;
  }
}
