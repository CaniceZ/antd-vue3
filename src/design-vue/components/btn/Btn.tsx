import { createVNode, defineComponent, type PropType } from 'vue'
import Modal from '../../plugins/modal'
import {
  Button as AButton,
  type ButtonProps as AButtonProps,
  type ModalFuncProps
} from 'ant-design-vue'
import { QuestionCircleOutlined } from '@ant-design/icons-vue'
import { GlobalComponentConstructor, PropsType } from '../../types/ts-helpers'
import { AButtonSlots } from './types'
import { getPrefix } from '../../_utils/common'

const [prefixName, prefixCls] = getPrefix('btn')

const btnProps = () => ({
  /** 按钮显示文本 */
  label: {
    type: String
  },
  /** 等于 `shape="circle"` */
  circle: {
    type: Boolean,
    default: false
  },
  /** 等于 `shape="round"` */
  round: {
    type: Boolean,
    default: false
  },
  /** 等于 `type="primary"` */
  primary: {
    type: Boolean,
    default: false
  },
  /** 等于 `type="dashed"` */
  dashed: {
    type: Boolean,
    default: false
  },
  /** 等于 `type="link"` */
  link: {
    type: Boolean,
    default: false
  },
  /** 等于 `type="text"` */
  text: {
    type: Boolean,
    default: false
  },
  /** 等于 `size="small"` */
  small: {
    type: Boolean,
    default: false
  },
  /** 等于 `size="large"` */
  large: {
    type: Boolean,
    default: false
  },
  /** 启用form表单的提交按钮功能，等于 `<button type="submit" />` */
  submit: {
    type: Boolean,
    default: false
  },
  /** 启用弹出确认框，点击确认才会执行 onClick 事件，可设置弹出框配置 */
  confirm: {
    type: [Boolean, Object] as PropType<boolean | ModalFuncProps>,
    default: false
  }
  // TODO feat: 是否考虑再添加一个 popconfirm 的属性
})

const Btn = defineComponent({
  name: prefixName,
  inheritAttrs: false,
  props: btnProps(),
  setup(props, { attrs, slots }) {
    function onConfirmClick(...args) {
      const defaultConfirm: ModalFuncProps = {
        title: `${props.label}确认`,
        content: `是否确定${props.label}？`,
        icon: createVNode(QuestionCircleOutlined),
        okButtonProps: {
          danger: attrs['danger'] === '' || attrs['danger'] === true
        },
        okText: '确定',
        cancelText: '取消',
        onOk() {
          const fn = attrs['onClick']
          if (typeof fn === 'function') {
            fn(...args)
          }
        }
      }
      if (typeof props.confirm === 'object') {
        Object.assign(defaultConfirm, props.confirm)
      }
      Modal.confirm(defaultConfirm)
    }

    return () => (
      <AButton
        class={prefixCls}
        type={
          props.primary
            ? 'primary'
            : props.dashed
            ? 'dashed'
            : props.link
            ? 'link'
            : props.text
            ? 'text'
            : 'default'
        }
        size={props.small ? 'small' : props.large ? 'large' : 'middle'}
        shape={props.circle ? 'circle' : props.round ? 'round' : void 0}
        html-type={props.submit ? 'submit' : void 0}
        {...(props.confirm ? { ...attrs, onClick: onConfirmClick } : attrs)}
        v-slots={slots}
      >
        {props.label}
      </AButton>
    )
  }
})

export type BtnProps = PropsType<AButtonProps> &
  Omit<PropsType<typeof Btn>, keyof PropsType<typeof btnProps>> &
  PropsType<typeof btnProps>
export type BtnExpose = {}
export type BtnSlots = Partial<AButtonSlots>

export default Btn as GlobalComponentConstructor<BtnProps, BtnSlots>
