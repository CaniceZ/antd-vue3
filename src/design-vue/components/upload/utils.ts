export const extname = (url = '') => {
  const temp = url.split('/')
  const filename = temp[temp.length - 1]
  const filenameWithoutSuffix = filename.split(/#|\?/)[0]
  return (/\.[^./\\]*$/.exec(filenameWithoutSuffix) || [''])[0]
}

const isImageFileType = type => !!type && type.indexOf('image/') === 0

export const isImageUrl = file => {
  if (isImageFileType(file.type)) {
    return true
  }
  const url = file.thumbUrl || file.url
  const extension = extname(url)
  if (
    /^data:image\//.test(url) ||
    /(webp|svg|png|gif|jpg|jpeg|jfif|bmp|dpg|ico)$/i.test(extension)
  ) {
    return true
  }
  if (/^data:/.test(url)) {
    // other file types of base64
    return false
  }
  if (extension) {
    // other file types which have extension
    return false
  }
  return true
}

export const getFileType = file => {
  if (!file) {
    return
  }
  const attachmentUrl = file.url

  const type = extname(attachmentUrl)
  let fileType = ''
  if (
    isImageFileType(file.type) ||
    /(webp|svg|png|gif|jpg|jpeg|jfif|bmp|dpg|ico)$/i.test(type)
  ) {
    fileType = 'image'
  } else if (type === '.pdf') {
    fileType = 'pdf'
  } else if (['.doc', '.dot'].includes(type)) {
    fileType = 'word'
  } else if (['.xls', '.xlsx'].includes(type)) {
    fileType = 'excel'
  } else {
    fileType = 'file'
  }
  return fileType
}

export const imageDimensions = file =>
  new Promise((resolve, reject) => {
    const img = new Image()

    img.onload = () => {
      const { naturalWidth: width, naturalHeight: height } = img
      resolve({ width, height })
    }

    img.onerror = () => {
      reject('There was some problem with the image.')
    }

    img.src = URL.createObjectURL(file)
  })

export const isUrl = (url: string) => {
  const regex = /^(http|https)/
  return regex.test(url.toLowerCase())
}
