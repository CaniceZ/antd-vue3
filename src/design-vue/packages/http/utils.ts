const toString = Object.prototype.toString

export function is(val: unknown, type: string) {
  return toString.call(val) === `[object ${type}]`
}

export function isString(val: unknown): val is string {
  return is(val, 'String')
}

export function isObject(val: any): val is Record<any, any> {
  return val !== null && is(val, 'Object')
}

export function isFunction(val: unknown): val is Function {
  return typeof val === 'function'
}

/**
 * Add the object as a parameter to the URL
 * @param baseUrl url
 * @param obj
 * @returns {string}
 * eg:
 *  let obj = {a: '3', b: '4'}
 *  setObjToUrlParams('www.baidu.com', obj)
 *  ==>www.baidu.com?a=3&b=4
 */
export function setObjToUrlParams(baseUrl: string, obj: any): string {
  let parameters = ''
  for (const key in obj) {
    parameters += key + '=' + encodeURIComponent(obj[key]) + '&'
  }
  parameters = parameters.replace(/&$/, '')
  return /\?$/.test(baseUrl)
    ? baseUrl + parameters
    : baseUrl.replace(/\/?$/, '?') + parameters
}

export function deepMerge<T = any>(src: any = {}, target: any = {}): T {
  let key: string
  for (key in target) {
    src[key] = isObject(src[key])
      ? deepMerge(src[key], target[key])
      : (src[key] = target[key])
  }
  return src
}

declare global {
  interface Navigator {
    msSaveBlob(blob: Blob, string): void
  }
}

/**
 * 根据后端接口文件流下载
 * @param {*} data
 * @param {*} filename
 * @param {*} mime
 * @param {*} bom
 */
export function downloadByData(
  data: BlobPart,
  filename: string,
  mime?: string,
  bom?: BlobPart
) {
  const blobData = typeof bom !== 'undefined' ? [bom, data] : [data]
  const blob = new Blob(blobData, { type: mime || 'application/octet-stream' })
  // 如果浏览器的msSaveBlob存在
  if (typeof window.navigator.msSaveBlob !== 'undefined') {
    // 使用msSaveBlob下载文件
    window.navigator.msSaveBlob(blob, filename)
  } else {
    // 创建一个URL
    const blobURL = window.URL.createObjectURL(blob)
    const tempLink = document.createElement('a')
    tempLink.style.display = 'none'
    tempLink.href = blobURL
    tempLink.setAttribute('download', filename)
    if (typeof tempLink.download === 'undefined') {
      tempLink.setAttribute('target', '_blank')
    }
    document.body.appendChild(tempLink)
    // 手动触发点击事件
    tempLink.click()
    // 点击完成后移除链接
    document.body.removeChild(tempLink)
    // 移出URL
    window.URL.revokeObjectURL(blobURL)
  }
}
