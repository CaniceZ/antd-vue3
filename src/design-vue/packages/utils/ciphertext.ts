import get from 'lodash-es/get'

type Options = {
  sign: '*' | string
  suffix: '_encrypt' | string
}

const defaultOptions: Options = {
  sign: '*',
  suffix: '_encrypt'
}

/**
 * 获取密文的密钥字段名
 * @param name
 * @param opts
 * @returns
 */
export function getCiphertextField(
  name: string,
  opts: Partial<Options> = defaultOptions
): string {
  const options = {
    ...defaultOptions,
    ...opts
  }
  return `${name}${options.suffix}`
}

/**
 * 获取密文对象实例
 * @param model 数据对象
 * @param name 密文字段名称 / 密文字段路径
 * @param opts.sign 脱敏显示符号
 * @param opts.suffix 脱敏密文字段名后缀
 * @returns
 */
export function getCiphertext(
  model: object,
  name: string,
  opts: Partial<Options> = defaultOptions
): {
  text: string
  cipherText: string
} {
  const options = {
    ...defaultOptions,
    ...opts
  }
  const emptyRes = {
    text: '',
    cipherText: ''
  }
  if (typeof model === 'object') {
    const text = get(model, name)
    const cipherText = get(model, getCiphertextField(name, options))
    if (text && cipherText && text.includes(options.sign)) {
      return {
        text,
        cipherText
      }
    } else {
      return emptyRes
    }
  } else {
    return emptyRes
  }
}

/**
 * 判断是否密文
 * @param model 数据对象
 * @param name 密文字段名称 / 密文字段路径
 * @param opts.sign 脱敏显示符号
 * @param opts.suffix 脱敏密文字段名后缀
 */
export function isCiphertext(
  model: object,
  name: string,
  opts: Partial<Options> = defaultOptions
): boolean {
  const options = {
    ...defaultOptions,
    ...opts
  }
  const ciphertext = getCiphertext(model, name, options)
  return Boolean(ciphertext.text && ciphertext.cipherText)
}
