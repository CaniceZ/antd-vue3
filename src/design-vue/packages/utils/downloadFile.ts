/**
 * 下载文件
 * @param url 下载路径
 * @returns void
 */
export default function downloadFile(url: string) {
  if (!url) return
  const a = document.createElement('a')
  a.href = url
  a.click()
}
