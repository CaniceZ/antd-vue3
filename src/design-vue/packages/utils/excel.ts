import { Message, Modal, YImportResultModal } from '../../'

/**
 * 导入文件
 * @param params
 * @param callback
 * @returns Promise
 */
export const importFile = async (params: {
  file: File | any
  title: string
  getBatchNumber: Function
  getImportResult: Function
  wrapCount?: number
  extraData?: Object
  templateCode?: string
  resultProps?: any
  resultSlots?: any
}) => {
  const {
    file,
    getBatchNumber,
    getImportResult,
    title,
    extraData,
    templateCode,
    resultProps,
    resultSlots,
    wrapCount
  } = params

  const formData = new FormData()
  formData.append('file', file)
  templateCode && formData.append('templateCode', templateCode) // 模板编码

  if (extraData) {
    Object.keys(extraData).forEach(key => {
      formData.append(key, extraData![key])
    })
  }

  const res: any = await getBatchNumber(formData)
  return new Promise((resolve, reject) => {
    const batchNumber = res?.batchNumber || res
    if (batchNumber) {
      Message.success('上传成功，稍后通知您导入结果!')
      importResultPolling({
        resultProps,
        resultSlots,
        getImportResult,
        batchNumber,
        title,
        wrapCount,
        doneCallback: () => {
          resolve(params)
        },
        timeoutCallback: () => {
          reject()
        }
      })
    } else {
      Message.error('上传失败，请稍后重试')
      reject()
    }
  })
}

/**
 * 导入结果轮询
 * @param params 轮询的请求参数
 * @param callback 成功回调
 */
export const importResultPolling = (params: {
  title: String
  batchNumber: string
  getImportResult: Function
  resultProps?: any
  resultSlots?: any
  wrapCount?: number
  doneCallback?: (...args: any) => void
  timeoutCallback?: (...args: any) => void
}) => {
  const {
    wrapCount,
    batchNumber,
    title,
    doneCallback,
    timeoutCallback,
    resultProps,
    resultSlots,
    getImportResult
  } = params
  if (!batchNumber) return
  function run(count = 1) {
    setTimeout(() => {
      let params = { batchNumber }
      if (resultProps?.extraData) {
        params = { ...params, ...resultProps.extraData }
      }
      const typeMap = resultProps?.typeMap
      getImportResult(params).then(
        res => {
          if (resultProps?.isPms) {
            // pms特殊导入结果处理
            if (res && res.id) {
              if (!resultProps?.hideModal) {
                Modal.create(YImportResultModal, {
                  list: [res],
                  typeMap,
                  importBizSet: [resultProps.importBiz],
                  title,
                  ...(resultProps || {})
                })
              }
              doneCallback?.({ res, typeMap, title })
            } else {
              if (count < (wrapCount || 20)) {
                run(++count)
              } else {
                Message.error('稍后请在最近导入结果中查看')
                timeoutCallback?.()
              }
            }
          } else {
            // excel通用
            const { status, msg } = res
            if (status === 4) {
              if (count < (wrapCount || 20)) {
                run(++count)
              } else {
                Message.error(msg || '稍后请在最近导入结果中查看')
                timeoutCallback?.()
              }
            } else {
              if (resultProps?.callback) {
                resultProps.callback(res)
              } else {
                if (!resultProps?.hideModal) {
                  Modal.create(
                    YImportResultModal,
                    {
                      title,
                      list: [res],
                      ...(resultProps || {})
                    },
                    resultSlots || {}
                  )
                }
                doneCallback?.({ res, typeMap, title })
              }
            }
          }
        },
        err => {
          Message.error(err?.msg || '上传失败，请稍后重试')
        }
      )
    }, 3000)
  }
  run()
}

// pms导出结果轮询
export const pmsExportPolling = (option: {
  batchNumber: string
  exportBiz: String
  userCode: String
  getPmsExportDetail: Function
  timeout?: (...args: any) => void
  done?: (...args: any) => void
  error?: (...args: any) => void
}) => {
  const {
    batchNumber,
    exportBiz,
    userCode,
    getPmsExportDetail,
    timeout,
    done,
    error
  } = option
  if (!batchNumber) return
  const params = {
    batchNumber: batchNumber,
    exportBiz,
    userCode
  }
  function run(count = 1) {
    setTimeout(() => {
      getPmsExportDetail(params).then(
        res => {
          const { code, data, msg } = res
          if (code === 0) {
            if (data && data.list.length > 0) {
              done?.(data)
              data.list &&
                data.list.forEach(item => {
                  if (item.excelUrl) {
                    window.location.href = item.excelUrl
                  } else {
                    Message.error(item.msg)
                  }
                })
            } else {
              if (count < 40) {
                run(++count)
              } else {
                Message.error('导出超时，请重新导出')
                timeout?.()
              }
            }
          } else {
            Message.error(msg)
            error?.()
          }
        },
        () => {
          error?.()
        }
      )
    }, 3000)
  }
  run()
}
