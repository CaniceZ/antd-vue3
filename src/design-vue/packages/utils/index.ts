import omit from './omit'
import pick from './pick'
import downloadFile from './downloadFile'
import { isCiphertext, getCiphertext, getCiphertextField } from './ciphertext'
import { needCLodop, loadCLodop, getLodop } from './LodopFuncs'
import { importFile, importResultPolling } from './excel'
import filterUnit from './filterUnit'
import filterMoney from './filterMoney'

export {
  omit,
  pick,
  downloadFile,
  isCiphertext,
  getCiphertext,
  getCiphertextField,
  needCLodop,
  loadCLodop,
  getLodop,
  importFile,
  importResultPolling,
  filterUnit,
  filterMoney
}
