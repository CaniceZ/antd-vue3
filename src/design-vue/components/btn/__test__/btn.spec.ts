import { mount } from '@vue/test-utils'
import { asyncExpect } from '../../../tests/utils'
import Btn from '..'
import Ygpd from '../../../index'

describe('YBtn', () => {
  it('Btn component type is ok?', async () => {
    const wrapper = mount(Btn as any, {
      props: {
        label: '这是btn',
        primary: true
      },
      shallow: true
    })

    // 断言 直接传入类型的值是否生效，如果生效则能获取type，条件成立则为true
    expect(wrapper.find('a-button-stub').attributes('type')).toBe('primary')

    await wrapper.setProps({ type: 'link' })

    // 断言 传入type，是否能够覆盖本身的直接传入的type值，如果能够获取属性type为link则证明可以覆盖，
    expect(wrapper.find('a-button-stub').attributes('type')).toBe('link')

    await wrapper.setProps({ submit: true })

    // 断言传入submit，如果htmltype为submit则证明修改成功，条件成立
    expect(wrapper.find('a-button-stub').attributes('htmltype')).toBe('submit')

    await wrapper.setProps({ circle: true })

    // 断言 设置按钮形状，如果在组件身上的属性shape值为circle则证明透传成功，条件成立
    expect(wrapper.find('a-button-stub').attributes('shape')).toBe('circle')

    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot()
    }, 0)
  })

  // it('Btn component confirm is ok?', async () => {
  //   const wrapper = mount(Btn as any, {
  //     props: {
  //       label: '这是btn',
  //       confirm: true,
  //     },
  //     global: {
  //       plugins: [Ygpd],
  //     },
  //   })

  //   await wrapper.find('.ygp-btn').trigger('click')
  //   await asyncExpect(() => {
  //     console.log(wrapper.html())
  //     // expect(wrapper.html()).toMatchSnapshot()
  //   }, 0)
  // })
})
