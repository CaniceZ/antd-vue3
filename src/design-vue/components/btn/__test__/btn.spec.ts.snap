// Vitest Snapshot v1

exports[`YBtn > Btn component type is ok? 1`] = `"<a-button-stub size=\\"middle\\" type=\\"link\\" htmltype=\\"submit\\" loading=\\"false\\" class=\\"ygp-btn\\" shape=\\"circle\\"></a-button-stub>"`;
